app/services/user-agent-service/datomstorage-sqlite.js

```diff
@@ -0,0 +1,161 @@
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+
+import {
+  datascript,
+  mori,
+  helpers,
+} from 'datascript-mori';
+
+import * as datomstore from './datomstore';
+import { DatomStorageSchemaV1 } from './datomstorage-schema';
+
+const { vector } = mori;
+const { DB_ADD, DB_RETRACT } = helpers;
+
+const d = datascript.core;
+const djs = datascript.js;
+
+export class DatomStorageSqlite {
+  constructor(db) {
+    this.db = db;
+  }
+
+  async init(sqlSchema = new DatomStorageSchemaV1()) {
+    await this.db.exec('PRAGMA foreign_keys = ON');
+    const v = await sqlSchema.createOrUpdate(this);
+
+    if (v !== sqlSchema.version) {
+      throw new Error(`Incorrect version ${v}; expected ${sqlSchema.version}.`);
+    }
+  }
+
+  async close() {
+    this.db.close();
+    return true;
+  }
+
+  async userVersion() {
+    return await this.db
+      .get('PRAGMA user_version')
+      .then((row) => Promise.resolve(row.user_version));
+  }
+
+  async listen(report) {
+    const txData = mori.get(report, mori.keyword('tx-data'));
+
+    await this.db.inTransaction(() => {
+      // Write 150 5-tuples at a time.  SQLite has a limit of 999 bound variables per statement.
+      let ps = mori.map((txChunk) => {
+        const qs = `(?, ?, ?, ?, ?)${', (?, ?, ?, ?, ?)'.repeat(mori.count(txChunk) - 1)}`;
+        return this.db
+          .run(`INSERT INTO transactions (e, a, v, tx, added) VALUES ${qs}`,
+               mori.intoArray(mori.mapcat((tx) => mori.vector(tx.e,
+                                                              tx.a,
+                                                              tx.v,
+                                                              tx.tx,
+                                                              tx.added ? 1 : 0), txChunk)));
+      }, mori.partitionAll(150, txData));
+
+      // ps is a lazy sequence.  intoArray forces, like `seq`.
+      ps = mori.intoArray(ps);
+      return Promise.all(mori.intoArray(ps));
+    });
+  }
+
+  async getLastTxInSnapshot() {
+    const [{ lastTx }] =
+      await this.db.all('SELECT COALESCE(MAX(tx), -1) AS lastTx FROM datoms');
+    return lastTx;
+  }
+
+  /**
+   * Replay transactions out of the DB against the given connection.
+   *
+   * @param fromTx smallest tx to replay.
+   * @param toTx largest tx to replay.
+   */
+  async replayTransactions(conn, fromTx = 0, toTx = 1 << 30) {
+    // TODO: larger upper bound?
+    const rows = await this.db.all(`
+      SELECT e, a, v, tx, added FROM transactions
+      WHERE tx BETWEEN ? AND ?
+      ORDER BY tx ASC`, [fromTx, toTx]);
+    mori.each(mori.partitionBy((row) => row.tx, rows), (txRows) => {
+      const assertions = mori.apply(vector, mori.map(
+        (txRow) => vector(txRow.added ? DB_ADD : DB_RETRACT, txRow.e, txRow.a, txRow.v, txRow.tx),
```

Comment:
I have a sneaking suspicion that this won't handle `:db/txInstant` correctly — it'll probably persist and replay the meta-addition itself, but perhaps it'll apply to a transaction that's not this one.